src/media_import/pathlike/apkg.py

```diff
@@ -0,0 +1,90 @@
+import json
+import zipfile
+from functools import cached_property
+from hashlib import md5
+from pathlib import Path
+from typing import Dict, List, Union
+
+from .base import FileLike, RootPath
+from .errors import (IncompatibleApkgFormatError, IsADirectoryError,
+                     MalformedURLError, RootNotFoundError)
+
+
+class ApkgRoot(RootPath):
+    raw: str
+    name: str
+    files: List["FileLike"]
+    path: Path
+    zip_file: zipfile.ZipFile
+
+    def __init__(self, path: Union[str, Path]) -> None:
+        self.raw = str(path)
+        try:
+            if isinstance(path, str):
+                self.path = Path(path)
+            else:
+                self.path = path
+            if not self.path.is_file():
+                if self.path.is_dir():
+                    raise IsADirectoryError()
+                else:
+                    raise RootNotFoundError()
+            if not self.path.suffix == ".apkg":
+                raise MalformedURLError()
+        except OSError:
+            raise MalformedURLError()
+        self.name = self.path.name
+        self.zip_file = zipfile.ZipFile(self.path, "r")
+        self.files = self.list_files()
+
+    def list_files(self) -> List["FileLike"]:
+        # The media file contains a mapping from media filenames inside the zip file to the original filenames.
+        old_to_new_name = self._media_dict()
+        files: List["FileLike"] = [
+            FileInZip(new, zip_path=self.path, name_in_zip=old)
+            for old, new in old_to_new_name.items()
+        ]
+        return files
```

Comment:
@RisingOrange , I think this is going to need to be a generator.  We need to test this with the AnKing deck that it will mostly be used for (the apkg linked to here:  https://www.reddit.com/r/medicalschoolanki/comments/swjqzh/anking_overhaul_step_1step_2_version_11_update/)

I tested it with the apkg there and got the following (the dialogue just hangs with the "Calculating number of files..." message.)

```
Qt warning: QLayout: Attempting to add QLayout "" to LocalTab "", which already has a layout 
Qt warning: QLayout: Attempting to add QLayout "" to ApkgTab "", which already has a layout 
Qt warning: QLayout: Attempting to add QLayout "" to GDriveTab "", which already has a layout 
Qt warning: QLayout: Attempting to add QLayout "" to MegaTab "", which already has a layout 
2023-06-27 13:59:59.992 python[16461:23706169] +[CATransaction synchronize] called within transaction
INFO 2023-06-27 14:04:41,458 debug 16461 13290864640 Stack trace (db.begin):
  File "/Users/andrewsanchez/.vscode/extensions/ms-python.python-2023.10.1/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydev_bundle/pydev_monkey.py", line 1118, in __call__
    ret = self.original_func(*self.args, **self.kwargs)

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/threading.py", line 930, in _bootstrap
    self._bootstrap_inner()

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/threading.py", line 973, in _bootstrap_inner
    self.run()

  File "/tmp/Anki2/addons21/ankihub/lib/sentry_sdk/integrations/threading.py", line 67, in run
    return old_run_func(self, *a, **kw)

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/concurrent/futures/thread.py", line 83, in _worker
    work_item.run()

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/site-packages/aqt/operations/__init__.py", line 239, in wrapped_op
    return self._op(mw.col)

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/site-packages/aqt/main.py", line 1427, in backup
    return col.create_backup(

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/site-packages/anki/collection.py", line 374, in create_backup
    self.db.begin()

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/site-packages/decorator.py", line 232, in fun
    return caller(func, *(extras + args), **kw)

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/site-packages/anki/hooks.py", line 89, in decorator_wrapper
    return repl(*args, **kwargs)

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/site-packages/anki/hooks.py", line 83, in repl
    new(*args, **kwargs)

  File "/tmp/Anki2/addons21/ankihub/debug.py", line 36, in <lambda>
    lambda *args, **kwargs: _log_stack("db.begin"),

  File "/tmp/Anki2/addons21/ankihub/debug.py", line 17, in _log_stack
    stack_trace = "\n".join(traceback.format_stack())

1687900181.189 GET /_anki/legacyPageDataINFO 2023-06-27 14:09:41,190 debug 16461 13290864640 Stack trace (db.begin):
  File "/Users/andrewsanchez/.vscode/extensions/ms-python.python-2023.10.1/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydev_bundle/pydev_monkey.py", line 1118, in __call__
    ret = self.original_func(*self.args, **self.kwargs)

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/threading.py", line 930, in _bootstrap
    self._bootstrap_inner()

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/threading.py", line 973, in _bootstrap_inner
    self.run()

  File "/tmp/Anki2/addons21/ankihub/lib/sentry_sdk/integrations/threading.py", line 67, in run
    return old_run_func(self, *a, **kw)

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/concurrent/futures/thread.py", line 83, in _worker
    work_item.run()

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/site-packages/aqt/operations/__init__.py", line 239, in wrapped_op
    return self._op(mw.col)

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/site-packages/aqt/main.py", line 1427, in backup
    return col.create_backup(

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/site-packages/anki/collection.py", line 374, in create_backup
    self.db.begin()

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/site-packages/decorator.py", line 232, in fun
    return caller(func, *(extras + args), **kw)

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/site-packages/anki/hooks.py", line 89, in decorator_wrapper
    return repl(*args, **kwargs)

  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/site-packages/anki/hooks.py", line 83, in repl
    new(*args, **kwargs)

  File "/tmp/Anki2/addons21/ankihub/debug.py", line 36, in <lambda>
    lambda *args, **kwargs: _log_stack("db.begin"),

  File "/tmp/Anki2/addons21/ankihub/debug.py", line 17, in _log_stack
    stack_trace = "\n".join(traceback.format_stack())
    1687900181.261 GET /_anki/legacyPageData
1687900182.079 GET /_anki/css/webview.css
1687900182.109 GET /_anki/css/deckbrowser.css
1687900182.128 GET /_anki/js/webview.js
Exception in thread 1687900413.847 GET /_anki/js/vendor/jquery.min.js
Thread-6:
Traceback (most recent call last):
  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/threading.py", line 973, in _bootstrap_inner
    self.run()
  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/site-packages/aqt/mediasrv.py", line 98, in run
    self.server.run()
  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/site-packages/waitress/server.py", line 322, in run
    self.asyncore.loop(
  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/site-packages/waitress/wasyncore.py", line 245, in loop
    poll_fun(timeout, map)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/site-packages/waitress/wasyncore.py", line 172, in poll
    r, w, e = select.select(r, w, e, timeout)
ValueError: filedescriptor out of range in select()
Exception in threading.excepthook:
Exception ignored in thread started by: <_pydev_bundle.pydev_monkey._NewThreadStartupWithTrace object at 0x7fd270ce7880>
Traceback (most recent call last):
  File "/Users/andrewsanchez/.vscode/extensions/ms-python.python-2023.10.1/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydev_bundle/pydev_monkey.py", line 1118, in __call__
    ret = self.original_func(*self.args, **self.kwargs)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/threading.py", line 930, in _bootstrap
    self._bootstrap_inner()
  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/threading.py", line 975, in _bootstrap_inner
    self._invoke_excepthook(self)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/ankihub_addon/lib/python3.9/threading.py", line 1257, in invoke_excepthook
    local_print("Exception in threading.excepthook:",
AttributeError: 'ErrorHandler' object has no attribute 'flush'
Exception ignored in sys.unraisablehook: <built-in function unraisablehook>
AttributeError: 'ErrorHandler' object has no attribute 'flush'
```

Reply:
I had to [refactor the tests](https://github.com/ankipalace/anki-media-import/pull/18/commits/13a73ff3dcb867b0318c60223dea7f3f73f96d5a) because they stopped working when more background tasks were added to the import workflow.